statistics/handle/lru_cache.go

```diff
@@ -0,0 +1,253 @@
+	l.Unlock()
+	if hit {
+		metrics.StatsCacheLRUCounter.WithLabelValues(typHit).Inc()
+	} else {
+		metrics.StatsCacheLRUCounter.WithLabelValues(typMiss).Inc()
+	}
+	return r, hit
+}
+
+func (l *internalLRUCache) get(key int64) (*statistics.Table, bool) {
+	element, exists := l.elements[key]
+	if !exists {
+		return nil, false
+	}
+	l.cache.MoveToFront(element)
+	return element.Value.(*cacheItem).value, true
+}
+
+// Put puts the (key, value) pair into the LRU Cache.
+func (l *internalLRUCache) Put(key int64, value *statistics.Table) {
+	var updated bool
+	l.Lock()
+	updated = l.put(key, value, value.MemoryUsage(), true)
+	l.Unlock()
+	if updated {
+		metrics.StatsCacheLRUCounter.WithLabelValues(typUpdate).Inc()
+		return
+	}
+	metrics.StatsCacheLRUCounter.WithLabelValues(typInsert).Inc()
+}
+
+func (l *internalLRUCache) put(key int64, value *statistics.Table, cost int64, tryEvict bool) bool {
+	defer func() {
+		if tryEvict {
+			l.evictIfNeeded()
+		}
+	}()
+	element, exists := l.elements[key]
+	if exists {
+		oldCost := element.Value.(*cacheItem).cost
+		element.Value.(*cacheItem).value = value
+		element.Value.(*cacheItem).cost = cost
+		l.cache.MoveToFront(element)
+		l.cost += cost - oldCost
+		return true
+	}
+	newCacheEntry := &cacheItem{
+		key:   key,
+		value: value,
+		cost:  cost,
+	}
+	element = l.cache.PushFront(newCacheEntry)
+	l.elements[key] = element
+	l.cost += cost
+	return false
+}
+
+// Del deletes the key-value pair from the LRU Cache.
+func (l *internalLRUCache) Del(key int64) {
+	var del bool
+	l.Lock()
+	del = l.del(key)
+	l.Unlock()
+	if del {
+		metrics.StatsCacheLRUCounter.WithLabelValues(typDel).Inc()
+	}
+}
+
+func (l *internalLRUCache) del(key int64) bool {
+	element := l.elements[key]
+	if element == nil {
+		return false
+	}
+	l.cache.Remove(element)
+	delete(l.elements, key)
+	l.cost -= element.Value.(*cacheItem).cost
+	return true
+}
+
+// Cost returns the current cost
+func (l *internalLRUCache) Cost() int64 {
+	l.RLock()
+	defer l.RUnlock()
+	return l.cost
+}
+
+// Keys returns the current Keys
+func (l *internalLRUCache) Keys() []int64 {
+	l.RLock()
+	defer l.RUnlock()
+	r := make([]int64, 0)
+	for _, v := range l.elements {
+		r = append(r, v.Value.(*cacheItem).key)
+	}
+	return r
+}
+
+// Values returns the current Values
+func (l *internalLRUCache) Values() []*statistics.Table {
+	l.RLock()
+	defer l.RUnlock()
+	r := make([]*statistics.Table, 0)
+	for _, v := range l.elements {
+		r = append(r, v.Value.(*cacheItem).value)
+	}
+	return r
+}
+
+// Len returns the current length
+func (l *internalLRUCache) Len() int {
+	l.RLock()
+	defer l.RUnlock()
+	return len(l.elements)
+}
+
+// CalculateCost re-calculate the memory message
+func (l *internalLRUCache) CalculateCost() {
+	var cost int64
+	l.Lock()
+	for _, v := range l.elements {
+		item := v.Value.(*cacheItem)
+		oldCost := item.cost
+		newCost := item.value.MemoryUsage()
+		item.cost = newCost
+		l.cost += newCost - oldCost
+	}
+	l.evictIfNeeded()
```

Comment:
It's a surprise to see evict under "CalculateCost". What about renaming CalculateCost or separate evict call from it.